testcontainers/src/watchdog.rs

```diff
@@ -0,0 +1,50 @@
+use crate::{clients::Cli, core::Docker};
+use conquer_once::Lazy;
+use signal_hook::{
+    consts::{SIGINT, SIGQUIT, SIGTERM},
+    iterator::Signals,
+};
+use std::{collections::BTreeSet, sync::Mutex, thread};
+
+static WATCHDOG: Lazy<Mutex<Watchdog>> = Lazy::new(|| {
+    let watchdog = Mutex::default();
+    let mut signals =
+        Signals::new(&[SIGTERM, SIGINT, SIGQUIT]).expect("failed to register signal handler");
+    thread::spawn(move || {
+        let signal_docker = Cli::default();
+        for signal in &mut signals {
+            for container_id in WATCHDOG
+                .lock()
+                .map(|s| s.containers.clone())
+                .unwrap_or_default()
+            {
+                signal_docker.stop(&container_id);
+                signal_docker.rm(&container_id);
+            }
+            let _ = signal_hook::low_level::emulate_default_handler(signal);
+        }
+    });
+    watchdog
```

Comment:
I think this can be slightly re-ordered to be a bit more readable:

```suggestion
    thread::spawn(move || {
        let signal_docker = Cli::default();
	let mut signals = Signals::new(&[SIGTERM, SIGINT, SIGQUIT]).expect("failed to register signal handler");
    
        for signal in &mut signals {
            for container_id in WATCHDOG
                .lock()
                .map(|s| s.containers.clone())
                .unwrap_or_default()
            {
                signal_docker.stop(&container_id);
                signal_docker.rm(&container_id);
            }

            let _ = signal_hook::low_level::emulate_default_handler(signal);
        }
    });
    
    Mutex::new(Watchdog::default())
```